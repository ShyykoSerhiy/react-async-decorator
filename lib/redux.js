var create = require('./fetcher').create;
function createReduxFetcher(ACTION, KEY) {
    var api;

    function use(aApi) {
        api = aApi;
        return function(next) {
            return function(action) {
                return next(action);
            }
        };
    }

    function getApi() {
        if (!api) {
            throw new Error('add middleware to redux');
        }
        return api;
    }

    function createAccessor(name) {
        function clear() {
            getApi().dispatch({ type: ACTION, action: 'clear', name: name });
        }
        function get(key) {
            var state = getApi().getState();
            var data = state[KEY];
            var ptr = data ? data[key] : undefined;
            return (ptr && ptr[key]) ? ptr[key].d : undefined;
        }
        function set(key, val) {
            getApi().dispatch({ type: ACTION, action: 'set', key: key, value: val, name: name });
        }
        function has(key) {
            var state = getApi().getState()[KEY];
            var data = state[name];
            var ptr = data ? data[key] : undefined;
            return !!ptr;
        }
        return { clear: clear, set: set, has: has, get: get };
    }

    function reducer(state, action) {
        if (action.type === ACTION) {
            var nstate = Object.assign({}, state);
            if (action.action === 'clear') {
                if (nstate[KEY]) {
                    nstate[KEY][action.name] = {};
                }
            } else if (action.action === 'set') {
                var key = action.key;
                var name = action.name;
                var data = nstate[KEY] = Object.assign({}, nstate[KEY]);
                var ptr = data[name] = Object.assign({}, data[name]);
                ptr[key] = { d: action.value };
            }
            return nstate;
        }
        return state;
    }

    var createFetcher = create(createAccessor);
    return { use: use, reducer: reducer, createFetcher: createFetcher };
}

module.exports.createReduxFetcher = createReduxFetcher;
