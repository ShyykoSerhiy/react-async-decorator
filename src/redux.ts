import { create } from './fetcher';
import { MiddlewareAPI, Dispatch, IDataItem, IActionFetch, typeFetcherFn } from './interfaces';
import { IOptionStore, createReducer, IOptionReducer } from './reduxReducer';
export { MiddlewareAPI, Dispatch, IDataItem, IActionFetch };

export function initRedux<State>() {
	function createReduxFetcher(opt: IOptionStore) {
		let api: MiddlewareAPI;

		function use(aApi: MiddlewareAPI): void {
			api = aApi;
		}

		const store: MiddlewareAPI = {
			getState() {
				return api.getState();
			},
			dispatch(action: any) {
				return api.dispatch(action);
			},
		};

    // reducer initialize after create call
    let reducer: (state: State, act: IActionFetch) => State;

    // it's sync callback which call after create call
    function createStore(opt: IOptionReducer<any>): MiddlewareAPI {
      reducer = createReducer<State>(opt);
      return store;
    }

		const fn = create({ ...opt, createStore }) as any;
		return { use, reducer, createFetcher: fn as typeof typeFetcherFn };
	}
	return createReduxFetcher;
}
