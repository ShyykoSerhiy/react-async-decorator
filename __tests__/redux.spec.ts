import { createReduxFetcher } from '../index';
import { createStore, applyMiddleware, combineReducers, compose } from 'redux';


test('empty initial state', async () => {
    const conf = createReduxFetcher('ACTION', 'action');
    const store = createStore(conf.reducer, {
        action: { }
    });
    conf.use(store);
    const fetcher = conf.createFetcher(() => Promise.resolve(123));
    try {
        fetcher.get()
    } catch (e) {}
    await fetcher.awaitAll()
    const state = store.getState();
    expect(state).toMatchSnapshot();
});

test('predefined state shouldns overwrite', async () => {
    const conf = createReduxFetcher('ACTION', 'action');
    const initialState = {"action":{"0":{"":{"d":999}}}};
    const store = createStore(conf.reducer, initialState);
    conf.use(store);
    const fetcher = conf.createFetcher(() => Promise.resolve(123));
    try {
        fetcher.get()
    } catch (e) {}
    await fetcher.awaitAll()
    const state = store.getState();
    expect(state).toMatchSnapshot();
});
