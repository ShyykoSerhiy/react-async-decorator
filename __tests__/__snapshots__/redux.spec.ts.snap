// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty initial state 1`] = `
Object {
  "action": Object {
    "0": Object {
      "": Object {
        "d": 123,
      },
    },
  },
}
`;

exports[`predefined state shouldns overwrite 1`] = `
Object {
  "action": Object {
    "0": Object {
      "": Object {
        "d": 999,
      },
    },
  },
}
`;
